[The0nlyJuan] iP

The0nlyJuan opened the pull request:
### Krona Chatbot

> "Your tasks are safe with Krona. Never forget another task!" – Krona Team

Krona helps you manage tasks efficiently and effectively. It's designed to make tracking your to-do list a breeze.

- Easy to use

- Fast and efficient

- Customizable task management

All you need to do is:

1. Download Krona [here](https://github.com/The0nlyJuan/ip/tree/v0.1).
2. Run the program by double-clicking.
3. Start adding your tasks.
4. Let Krona manage them for you! :wink:

And it is **FREE**!

Features:

- [x] Managing tasks
- [x] Setting deadlines 
- [ ] Task reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

The0nlyJuan committed:
Rename Duke to Krona and added greetings and goodbye

The0nlyJuan committed:
Add command echo and exit

The0nlyJuan committed:
Add text storing and displaying

The0nlyJuan committed:
Add Task.java

The0nlyJuan committed:
Change Krona.java to handle Task.java usage

The0nlyJuan committed:
Add Deadline.java, Event.java, ToDo.java as subclasses of Task.java

The0nlyJuan committed:
Change Krona.java to handle new subclasses

The0nlyJuan committed:
Change runtest.sh and added expected text and passed

The0nlyJuan committed:
add empty description handling

The0nlyJuan committed:
Add error message for marking and unmarking invalid index numbers

The0nlyJuan committed:
add delete method and use ArrayList instead of fixed array

The0nlyJuan committed:
Add methods for saving data into file

The0nlyJuan committed:
Add LocalDateTime for Deadline and Event class

The0nlyJuan committed:
Add Ui.java

The0nlyJuan committed:
Add TaskList.java

The0nlyJuan committed:
Add rest of the needed files for OOP

The0nlyJuan committed:
Add missing commands and update Parser

The0nlyJuan committed:
Move all classes to corresponding packages

The0nlyJuan committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

The0nlyJuan committed:
Change Ui.java

The0nlyJuan committed:
Add 2 test classes for Deadline and Task

The0nlyJuan committed:
Add javadoc comments to most non-trivial and non-private classes/methods

The0nlyJuan committed:
Follow coding standard

The0nlyJuan committed:
Add FindCommand.java and update parser

The0nlyJuan committed:
Bug fix for empty event, deadline and todo

The0nlyJuan committed:
Add GUI for Krona ChatBot

The0nlyJuan committed:
Add CheckStyle Files and Correct all CheckStyle Issues

josh1248 commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,34 @@
+            AnchorPane ap = fxmlLoader.load();
+            Scene scene = new Scene(ap);
+            stage.setScene(scene);
+            stage.setMinHeight(220);
```
I believe this could be represented as a constant in the Main class, for example private static int MIN_HEIGHT_PX = 220. This would avoid the issue of magic constants, and explains the units involved in setting the height as well.

josh1248 commented on src/main/java/krona/command/AddCommand.java:
```diff
@@ -0,0 +1,44 @@
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws KronaException {
+        tasks.addTask(task);
+        StringBuilder output = new StringBuilder();
+        output.append("Got it. I've added this task:\n");
```
Would this belong in the purview of the UI as it is feedback to the user? Could consider.

josh1248 commented on src/main/java/krona/command/DeleteCommand.java:
```diff
@@ -0,0 +1,52 @@
+            output.append(task.toString()).append("\n");
+            output.append("Now you have ").append(tasks.size()).append(" tasks in the list.");
+
+            storage.save(tasks);
```
Love the use of line breaks here to separate the different 'phases' of the execute.

josh1248 commented on src/main/java/krona/parser/Parser.java:
```diff
@@ -0,0 +1,131 @@
+package krona.parser;
+
+import krona.command.AddCommand;
```
nice alphabetical order to easily identify the import.

josh1248 commented on src/main/java/krona/parser/Parser.java:
```diff
@@ -0,0 +1,131 @@
+     * @return The Command object corresponding to the user input.
+     */
+    public static Command parse(String input) {
+        String[] words = input.split(" ", 2);
```
this is not coding standard related, but i think this may detract from the project scope if the task is, for example, `todo borrow book` (example given from week 2 ip project requirements).

josh1248 commented on src/main/java/krona/task/Event.java:
```diff
@@ -0,0 +1,49 @@
+ */
+public class Event extends Task {
+    private static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d/M/yyyy HHmm");
+    protected LocalDateTime startDateTime;
```
since you have getter methods, it would be appropriate to place the access attributes here as `private`.

josh1248 commented on src/main/java/krona/storage/Storage.java:
```diff
@@ -0,0 +1,74 @@
+import krona.task.TaskList;
+
+
+/**
```
Very helpful class and method explanations here.

yimjianbing commented on src/main/java/krona/ui/Ui.java:
```diff
@@ -0,0 +1,71 @@
+        setCombinedMessage("Bye. Hope to see you again soon!");
+    }
+
+    public void exitApp() {
```
Really like this implementation to exit from the javafx application after the user input bye! Good job!

yimjianbing commented on src/main/resources/view/DialogBox.fxml:
```diff
@@ -0,0 +1,22 @@
+<?import javafx.scene.image.ImageView?>
+<?import javafx.scene.layout.HBox?>
+
+<fx:root alignment="TOP_RIGHT" maxHeight="1.7976931348623157E308" maxWidth="1.7976931348623157E308" prefWidth="400.0" stylesheets="@../css/dialog-box.css" type="javafx.scene.layout.HBox" xmlns="http://javafx.com/javafx/22" xmlns:fx="http://javafx.com/fxml/1">
```
Small nitpick but I don't think your .fxml is working? Your .fxml is stating it using java SDK vers 22, usually because every time you use scenebuilder to save the edits onto your .fxml file, it will set the version as xmlns="http://javafx.com/javafx/22"  instead of xmlns="http://javafx.com/javafx/17", so need to edit this file every time you use scenebuilder

The0nlyJuan replied:
Oh okay I did not even know this. Thank you!

yimjianbing commented on src/main/resources/view/MainWindow.fxml:
```diff
@@ -0,0 +1,19 @@
+<?import javafx.scene.layout.AnchorPane?>
+<?import javafx.scene.layout.VBox?>
+
+<AnchorPane maxHeight="-Infinity" maxWidth="-Infinity" minHeight="-Infinity" minWidth="-Infinity" prefHeight="600.0" prefWidth="400.0" stylesheets="@../css/main.css" xmlns="http://javafx.com/javafx/22" xmlns:fx="http://javafx.com/fxml/1" fx:controller="MainWindow">
```
same here with the java version being 22 instead of 17!

yimjianbing commented on src/test/java/krona/KronaTest.java:
```diff
@@ -0,0 +1,4 @@
+package krona;
+
+public class KronaTest {
```
I think having a JUnit test for the Krono app itself might be a bit impractical? Since you would have to isolate many of its dependent classes with SUTs and test drivers, but definitely not impossible

yimjianbing commented on src/main/java/krona/task/Deadline.java:
```diff
@@ -0,0 +1,40 @@
+ * can input date and time.
+ */
+public class Deadline extends Task {
+    private static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d/M/yyyy HHmm");
```
Maybe you can change the formatter name to be in all caps instead, since it has the final modifier, it means its a constant, so we have to put its name in all caps according to the coding standard

yimjianbing commented on src/main/java/krona/task/Event.java:
```diff
@@ -0,0 +1,49 @@
+ * and time to an end date and time.
+ */
+public class Event extends Task {
+    private static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d/M/yyyy HHmm");
```
Since this variable, formatter, is also final and thus a constant, maybe you should change its name to FORMATTER as well

yimjianbing commented on src/main/java/krona/command/MarkCommand.java:
```diff
@@ -0,0 +1,49 @@
+            Task task = tasks.get(taskIndex);
+            task.markDone();
+
+            StringBuilder output = new StringBuilder();
```
I really like the usage of StringBuilders here! Good job!

yimjianbing commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,34 @@
+ */
+public class Main extends Application {
+
+    private Krona krona = new Krona("data/tasks.txt");
```
I'm not sure if its my side, but when I used the Bye Command with your Krono, it was unable to save the tasks into a task.txt, and I couldn't find a tasks.txt in the directory, so maybe you can add it to your git? Unless I'm blind in which I sincerely apologise!

The0nlyJuan replied:
Hey! Am i supposed to add the data/tasks.txt to git? i just checked and it was under ignored files because i assume its supposed to be for each individual computer

yimjianbing reviewed (commented):
Overall the code is super clean and readable. I could easily navigate around the different classes, and the methods/variable names are all really well done. I knew what everything did, and I really enjoy how some of the features were implemented! Great job!

The0nlyJuan committed:
Extension for help command

User does not know the list of available commands.

Have a new command "help" to display list of commands

The0nlyJuan committed:
Change build.gradle to include archiveFileName

The0nlyJuan committed:
Add Assertions

The0nlyJuan committed:
Change some Code Quality issues

The0nlyJuan committed:
Merge pull request #1 from The0nlyJuan/branch-A-Assertions

Add Assertions

The0nlyJuan committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add Assertions

The0nlyJuan committed:
Merge pull request #2 from The0nlyJuan/branch-A-CodeQuality

Branch code quality

The0nlyJuan committed:
Add error handling for mark, unmark and delete without index

The0nlyJuan committed:
Bug fix for error handling of invalid date format in event and deadline class

The0nlyJuan committed:
Update User guide for readme in docs

The0nlyJuan committed:
Rename img.png to Ui.png

The0nlyJuan committed:
Change README.md

The0nlyJuan committed:
Change image name to display correct image

The0nlyJuan committed:
Add Background image to GUI

The0nlyJuan committed:
Tweak GUI

The0nlyJuan committed:
Change Welcome Message

The0nlyJuan committed:
Add more JUnit Test Cases

The0nlyJuan committed:
Change Ui image in docs

The0nlyJuan committed:
Update with new image

The0nlyJuan committed:
Change name of image

The0nlyJuan committed:
Finish User Guide

The0nlyJuan committed:
Fix hardcoded filepath issue
